File: python/09_变态跳台阶.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

class Solution():
    def jumpFloorII(self, n):
        """
        循环计算法
        """
        if n == 0:
            return 0
        
        tmp = 1
        for i in range(n - 1):
            tmp *= 2
        return tmp

    def jumpFloorII2(self, n):
        """
        直接计算法

        """
        if n == 0:
            return 0
        else:
            return 2**(n - 1)

    def jumpFloorII3(self, n):
        """
        一行方法
        """
        return 0 if n == 0 else 2**(n - 1)


def main():
    s = Solution()
    print(s.jumpFloorII(9))
    print(s.jumpFloorII2(9))
    print(s.jumpFloorII3(9))


if __name__ == "__main__":
    main()

File: cpp/09_变态跳台阶.cpp
#include <iostream>
#include <cmath>

class Solution {
public:
    // 常规解法
    int jumpFloorII(int n) {
        if (n == 0) {
            return n;
        }
        
        int tmp = 1;
        for (int i = 0; i < n - 1; i++) {
            tmp *= 2;
        }
        return tmp;
    }

    // 库函数解法
    int jumpFloorII2(int n) {
        if (n == 0) {
            return n;
        }

        return pow(2, n);
    }
};

int main(int argc, char* argv[])
{
    Solution s;
    std::cout << s.jumpFloorII(9) << std::endl;
    std::cout << s.jumpFloorII2(9) << std::endl;
    
    return 0;
}